### ISSUE TITLE
No `unbound variable` reported for variable self assignment inside function

### ISSUE THREAD

**Issue by ntjess** [microsoft/pylance-release#4634]
## Environment data

-   Language Server version: 2023.7.30
-   OS and version: Windows 10 21H2
-   Python version (& distribution if applicable, e.g. Anaconda): 3.10 mambaforge

## Code Snippet

```python
y = y

def test():
    x = x
    z

def test2():
    x.next()
```

## Expected behavior

The line `x = x` inside `test()` should show a ~`"x" is not defined`~ `local variable 'x' referenced before assignment` error.

*Edit: I realize after running the snippet that the error should be for an unbound local variable, not undefined variable.*

Looking at the [docs](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportUnboundVariable) I would assume this case should be properly identified

## Actual behavior
In all cases except `x = x`, Pylance shows the correct error. But in this case, it is treated as a valid assignment.

## Logs

<details>

```console
2023-07-20 13:29:17.610 [info] [Info  - 1:29:17 PM] (11248) Pylance language server 2023.7.30 (pyright 74d8f3c4) starting
2023-07-20 13:29:17.611 [info] [Info  - 1:29:17 PM] (11248) Server root directory: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist
2023-07-20 13:29:17.616 [info] [Info  - 1:29:17 PM] (11248) Starting service instance "<repo>"
2023-07-20 13:29:17.660 [info] (11248) No configuration file found.
2023-07-20 13:29:17.660 [info] (11248) pyproject.toml file found at <user-path>\<repo-path>.
2023-07-20 13:29:17.661 [info] [Info  - 1:29:17 PM] (11248) Setting pythonPath for service "<repo>": "<user-path>\mambaforge\envs\py310\python.exe"
2023-07-20 13:29:17.666 [info] [Info  - 1:29:17 PM] (11248) Loading pyproject.toml file at <user-path>\<repo-path>\pyproject.toml
2023-07-20 13:29:17.670 [info] [Info  - 1:29:17 PM] (11248) Pyproject file "<user-path>\<repo-path>\pyproject.toml" has no "[tool.pyright]" section.
2023-07-20 13:29:17.670 [info] [Warn  - 1:29:17 PM] (11248) stubPath <user-path>\<repo-path>\typings is not a valid directory.
2023-07-20 13:29:17.784 [info] [Info  - 1:29:17 PM] (11248) Assuming Python version 3.10
2023-07-20 13:29:17.784 [info] (11248) Assuming Python platform Windows
2023-07-20 13:29:17.972 [info] [Info  - 1:29:17 PM] (11248) Search paths for <user-path>\<repo-path>
2023-07-20 13:29:17.972 [info] [Info  - 1:29:17 PM] (11248)   <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\typeshed-fallback\stdlib
2023-07-20 13:29:17.973 [info] [Info  - 1:29:17 PM] (11248)   <user-path>\<repo-path>
2023-07-20 13:29:17.973 [info] [Info  - 1:29:17 PM] (11248)   <user-path>\<repo-path>\typings
2023-07-20 13:29:17.973 [info] [Info  - 1:29:17 PM] (11248)   <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\typeshed-fallback\stubs\...
2023-07-20 13:29:17.973 [info] [Info  - 1:29:17 PM] (11248)   <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\bundled\stubs
2023-07-20 13:29:17.974 [info] [Info  - 1:29:17 PM] (11248)   <user-path>\mambaforge\envs\py310\DLLs
2023-07-20 13:29:17.974 [info] [Info  - 1:29:17 PM] (11248)   <user-path>\mambaforge\envs\py310\Lib
2023-07-20 13:29:17.974 [info] [Info  - 1:29:17 PM] (11248)   <user-path>\mambaforge\envs\py310
2023-07-20 13:29:17.974 [info] [Info  - 1:29:17 PM] (11248)   <user-path>\mambaforge\envs\py310\Lib\site-packages
2023-07-20 13:29:17.974 [info] [Info  - 1:29:17 PM] (11248) Adding fs watcher for library directories:
 <user-path>\mambaforge\envs\py310\DLLs
<user-path>\mambaforge\envs\py310\lib
<user-path>\mambaforge\envs\py310
<user-path>\mambaforge\envs\py310\lib\site-packages
2023-07-20 13:29:17.976 [info] [Info  - 1:29:17 PM] (11248) Adding fs watcher for directories:
 <user-path>\<repo-path>
2023-07-20 13:29:17.976 [info] (11248) Searching for source files
2023-07-20 13:29:17.983 [info] [Info  - 1:29:17 PM] (11248) Found 8 source files
2023-07-20 13:29:18.108 [info] (11248) [FG] parsing: <user-path>\<repo-path>\<repo-file>.py (102ms)
2023-07-20 13:29:18.214 [info] (11248) [FG] parsing: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\typeshed-fallback\stdlib\builtins.pyi [fs read 5ms] (105ms)
2023-07-20 13:29:18.272 [info] (11248) [FG] binding: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\typeshed-fallback\stdlib\builtins.pyi (56ms)
2023-07-20 13:29:18.272 [info] (11248) [FG] binding: <user-path>\<repo-path>\<repo-file>.py (1ms)
2023-07-20 13:29:18.285 [info] [Info  - 1:29:18 PM] (11248) Background analysis(1) root directory: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist
2023-07-20 13:29:18.286 [info] [Info  - 1:29:18 PM] (11248) Background analysis(1) started
2023-07-20 13:29:18.286 [info] (11248) Background analysis message: setConfigOptions
2023-07-20 13:29:18.286 [info] (11248) Background analysis message: setImportResolver
2023-07-20 13:29:18.286 [info] (11248) Background analysis message: ensurePartialStubPackages
2023-07-20 13:29:18.392 [info] (11248) Background analysis message: setConfigOptions
2023-07-20 13:29:18.392 [info] (11248) Background analysis message: setTrackedFiles
2023-07-20 13:29:18.404 [info] (11248) Background analysis message: markAllFilesDirty
2023-07-20 13:29:18.404 [info] (11248) Background analysis message: setFileOpened
2023-07-20 13:29:18.405 [info] (11248) Background analysis message: getDiagnosticsForRange
2023-07-20 13:29:18.406 [info] (11248) Background analysis message: analyze
2023-07-20 13:29:18.469 [info] (11248) [BG(1)] analyzing: <user-path>\<repo-path>\<repo-file>.py ...
2023-07-20 13:29:18.469 [info] (11248) [BG(1)]   parsing: <user-path>\<repo-path>\<repo-file>.py (62ms)
2023-07-20 13:29:18.534 [info] [Info  - 1:29:18 PM] (11248) Indexer background runner(2) root directory: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist (index)
2023-07-20 13:29:18.534 [info] [Info  - 1:29:18 PM] (11248) Indexing(2) started
2023-07-20 13:29:18.566 [info] (11248) [BG(1)]   parsing: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\typeshed-fallback\stdlib\builtins.pyi [fs read 3ms] (96ms)
2023-07-20 13:29:18.611 [info] (11248) [BG(1)]   binding: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\typeshed-fallback\stdlib\builtins.pyi (44ms)
2023-07-20 13:29:18.612 [info] (11248) [BG(1)]   binding: <user-path>\<repo-path>\<repo-file>.py (1ms)
2023-07-20 13:29:18.664 [info] (11248) [BG(1)]   checking: <user-path>\<repo-path>\<repo-file>.py ...
2023-07-20 13:29:18.664 [info] (11248) [BG(1)]     parsing: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\typeshed-fallback\stdlib\typing.pyi [fs read 1ms] (31ms)
2023-07-20 13:29:18.683 [info] (11248) [BG(1)]     binding: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\typeshed-fallback\stdlib\typing.pyi (18ms)
2023-07-20 13:29:18.691 [info] (11248) [BG(1)]     parsing: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\typeshed-fallback\stdlib\typing_extensions.pyi [fs read 0ms] (4ms)
2023-07-20 13:29:18.695 [info] (11248) [BG(1)]     binding: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\typeshed-fallback\stdlib\typing_extensions.pyi (3ms)
2023-07-20 13:29:18.722 [info] (11248) [IDX(2)] scan packages <user-path>\<repo-path> ...
2023-07-20 13:29:18.723 [info] (11248) [IDX(2)]   read stdlib indices (185ms)
2023-07-20 13:29:18.726 [info] (11248) [BG(1)]     parsing: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\typeshed-fallback\stdlib\_typeshed\__init__.pyi [fs read 0ms] (6ms)
2023-07-20 13:29:18.731 [info] (11248) [BG(1)]     binding: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\typeshed-fallback\stdlib\_typeshed\__init__.pyi (4ms)
2023-07-20 13:29:18.744 [info] (11248) [BG(1)]     parsing: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\typeshed-fallback\stdlib\types.pyi [fs read 1ms] (13ms)
2023-07-20 13:29:18.750 [info] (11248) [BG(1)]     binding: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\typeshed-fallback\stdlib\types.pyi (5ms)
2023-07-20 13:29:18.758 [info] (11248) [BG(1)]     parsing: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\typeshed-fallback\stdlib\abc.pyi [fs read 0ms] (1ms)
2023-07-20 13:29:18.759 [info] (11248) [BG(1)]     binding: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\typeshed-fallback\stdlib\abc.pyi (1ms)
2023-07-20 13:29:18.766 [info] (11248) [BG(1)]   checking: <user-path>\<repo-path>\<repo-file>.py (154ms)
2023-07-20 13:29:18.766 [info] (11248) [BG(1)] analyzing: <user-path>\<repo-path>\<repo-file>.py (360ms)
2023-07-20 13:29:18.770 [info] (11248) Background analysis message: getSemanticTokens full
2023-07-20 13:29:18.771 [info] (11248) [BG(1)] getSemanticTokens full at <user-path>\<repo-path>\<repo-file>.py (3ms)
2023-07-20 13:29:18.772 [info] (11248) Background analysis message: getSemanticTokens range
2023-07-20 13:29:18.773 [info] (11248) [BG(1)] getSemanticTokens range 0:0 - 7:12 at <user-path>\<repo-path>\<repo-file>.py (1ms)
2023-07-20 13:29:18.773 [info] (11248) Background analysis message: getDiagnosticsForRange
2023-07-20 13:29:18.777 [info] (11248) Background analysis message: resumeAnalysis
2023-07-20 13:29:18.777 [info] (11248) [BG(1)] indexing: <user-path>\<repo-path>\<repo-file>.py ...
2023-07-20 13:29:18.778 [info] (11248) [BG(1)]   parsing: <user-path>\<repo-path>\<repo-file>.py [fs read 0ms] (0ms)
2023-07-20 13:29:18.778 [info] (11248) [BG(1)]   binding: <user-path>\<repo-path>\<repo-file>.py (1ms)
2023-07-20 13:29:18.778 [info] (11248) [BG(1)] indexing: <user-path>\<repo-path>\<repo-file>.py [found 1] (1ms)
2023-07-20 13:29:18.778 [info] (11248) Indexing Done: <user-path>\<repo-path>\<repo-file>.py
2023-07-20 13:29:18.829 [info] (11248) [BG(1)] indexing: <user-path>\<repo-path>\<repo-file>.py ...
2023-07-20 13:29:18.829 [info] (11248) [BG(1)]   parsing: <user-path>\<repo-path>\<repo-file>.py [fs read 1ms] (53ms)
2023-07-20 13:29:18.831 [info] (11248) [BG(1)]   binding: <user-path>\<repo-path>\<repo-file>.py (2ms)
2023-07-20 13:29:18.832 [info] (11248) [BG(1)] indexing: <user-path>\<repo-path>\<repo-file>.py [found 2] (56ms)
2023-07-20 13:29:18.832 [info] (11248) Indexing Done: <user-path>\<repo-path>\<repo-file>.py
2023-07-20 13:29:18.848 [info] (11248) [BG(1)] indexing: <user-path>\<repo-path>\<repo-file>.py ...
2023-07-20 13:29:18.848 [info] (11248) [BG(1)]   parsing: <user-path>\<repo-path>\<repo-file>.py [fs read 1ms] (16ms)
2023-07-20 13:29:18.883 [info] (11248) [BG(1)]   binding: <user-path>\<repo-path>\<repo-file>.py ...
2023-07-20 13:29:18.883 [info] (11248) [BG(1)]     parsing: <user-path>\<repo-path>\<repo-file>.py [fs read 0ms] (34ms)
2023-07-20 13:29:18.889 [info] (11248) [BG(1)]     binding: <user-path>\<repo-path>\<repo-file>.py (5ms)
2023-07-20 13:29:18.890 [info] (11248) [BG(1)]   binding: <user-path>\<repo-path>\<repo-file>.py (43ms)
2023-07-20 13:29:18.892 [info] (11248) [BG(1)] indexing: <user-path>\<repo-path>\<repo-file>.py [found 10] (60ms)
2023-07-20 13:29:18.892 [info] (11248) Indexing Done: <user-path>\<repo-path>\<repo-file>.py
2023-07-20 13:29:18.904 [info] (11248) [BG(1)] indexing: <user-path>\<repo-path>\<repo-file>.py ...
2023-07-20 13:29:18.904 [info] (11248) [BG(1)]   parsing: <user-path>\<repo-path>\<repo-file>.py [fs read 1ms] (12ms)
2023-07-20 13:29:18.905 [info] (11248) [BG(1)]   binding: <user-path>\<repo-path>\<repo-file>.py (2ms)
2023-07-20 13:29:18.907 [info] (11248) [BG(1)] indexing: <user-path>\<repo-path>\<repo-file>.py [found 8] (15ms)
2023-07-20 13:29:18.907 [info] (11248) Indexing Done: <user-path>\<repo-path>\<repo-file>.py
2023-07-20 13:29:18.907 [info] (11248) [BG(1)] indexing: <user-path>\<repo-path>\<repo-file>.py [found 3] (1ms)
2023-07-20 13:29:18.907 [info] (11248) Indexing Done: <user-path>\<repo-path>\<repo-file>.py
2023-07-20 13:29:18.910 [info] (11248) [BG(1)] indexing: <user-path>\<repo-path>\<repo-file>.py ...
2023-07-20 13:29:18.910 [info] (11248) [BG(1)]   parsing: <user-path>\<repo-path>\<repo-file>.py [fs read 0ms] (2ms)
2023-07-20 13:29:18.911 [info] (11248) [BG(1)]   binding: <user-path>\<repo-path>\<repo-file>.py (1ms)
2023-07-20 13:29:18.911 [info] (11248) [BG(1)] indexing: <user-path>\<repo-path>\<repo-file>.py [found 6] (4ms)
2023-07-20 13:29:18.912 [info] (11248) Indexing Done: <user-path>\<repo-path>\<repo-file>.py
2023-07-20 13:29:18.914 [info] (11248) [BG(1)] indexing: <user-path>\<repo-path>\<repo-file>.py ...
2023-07-20 13:29:18.915 [info] (11248) [BG(1)]   parsing: <user-path>\<repo-path>\<repo-file>.py [fs read 0ms] (3ms)
2023-07-20 13:29:18.915 [info] (11248) [BG(1)]   binding: <user-path>\<repo-path>\<repo-file>.py (0ms)
2023-07-20 13:29:18.915 [info] (11248) [BG(1)] indexing: <user-path>\<repo-path>\<repo-file>.py [found 4] (4ms)
2023-07-20 13:29:18.915 [info] (11248) Indexing Done: <user-path>\<repo-path>\<repo-file>.py
2023-07-20 13:29:18.919 [info] (11248) [BG(1)] indexing: <user-path>\<repo-path>\<repo-file>.py [found 16] (3ms)
2023-07-20 13:29:18.919 [info] (11248) Indexing Done: <user-path>\<repo-path>\<repo-file>.py
2023-07-20 13:29:19.036 [info] (11248) Background analysis message: getDiagnosticsForRange
2023-07-20 13:29:19.040 [info] (11248) Background analysis message: getDiagnosticsForRange
2023-07-20 13:29:19.280 [info] (11248) [IDX(2)] scan packages <user-path>\<repo-path> (743ms)
2023-07-20 13:29:19.281 [info] [Info  - 1:29:19 PM] (11248) scanned(2) 162 files over 1 exec env
2023-07-20 13:29:19.362 [info] (11248) [IDX(2)] index packages <user-path>\<repo-path> ...
2023-07-20 13:29:19.363 [info] (11248) [IDX(2)]   index execution environment <user-path>\<repo-path> ...
2023-07-20 13:29:19.363 [info] (11248) [IDX(2)]     indexing: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\bundled\stubs\matplotlib\_afm.pyi [skipped: private or protected file] (0ms)
2023-07-20 13:29:19.363 [info] (11248) [IDX(2)]     indexing: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\bundled\stubs\matplotlib\_cm.pyi [skipped: private or protected file] (0ms)
2023-07-20 13:29:19.363 [info] (11248) [IDX(2)]     indexing: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\bundled\stubs\matplotlib\_color_data.pyi [skipped: private or protected file] (1ms)
2023-07-20 13:29:19.364 [info] (11248) [IDX(2)]     indexing: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\bundled\stubs\matplotlib\_constrained_layout.pyi [skipped: private or protected file] (0ms)
2023-07-20 13:29:19.364 [info] (11248) [IDX(2)]     indexing: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\bundled\stubs\matplotlib\_c_internal_utils.pyi [skipped: private or protected file] (0ms)
2023-07-20 13:29:19.364 [info] (11248) [IDX(2)]     indexing: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\bundled\stubs\matplotlib\_docstring.pyi [skipped: private or protected file] (0ms)
2023-07-20 13:29:19.364 [info] (11248) [IDX(2)]     indexing: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\bundled\stubs\matplotlib\_enums.pyi [skipped: private or protected file] (0ms)
2023-07-20 13:29:19.364 [info] (11248) [IDX(2)]     indexing: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\bundled\stubs\matplotlib\_fontconfig_pattern.pyi [skipped: private or protected file] (0ms)
2023-07-20 13:29:19.364 [info] (11248) [IDX(2)]     indexing: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\bundled\stubs\matplotlib\_layoutgrid.pyi [skipped: private or protected file] (0ms)
2023-07-20 13:29:19.365 [info] (11248) [IDX(2)]     indexing: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\bundled\stubs\matplotlib\_mathtext.pyi [skipped: private or protected file] (0ms)
2023-07-20 13:29:19.365 [info] (11248) [IDX(2)]     indexing: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\bundled\stubs\matplotlib\_mathtext_data.pyi [skipped: private or protected file] (0ms)
2023-07-20 13:29:19.365 [info] (11248) [IDX(2)]     indexing: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\bundled\stubs\matplotlib\_pylab_helpers.pyi [skipped: private or protected file] (0ms)
2023-07-20 13:29:19.365 [info] (11248) [IDX(2)]     indexing: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\bundled\stubs\matplotlib\_text_helpers.pyi [skipped: private or protected file] (0ms)
2023-07-20 13:29:19.365 [info] (11248) [IDX(2)]     indexing: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\bundled\stubs\matplotlib\_tight_bbox.pyi [skipped: private or protected file] (0ms)
2023-07-20 13:29:19.366 [info] (11248) [IDX(2)]     indexing: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\bundled\stubs\matplotlib\_tight_layout.pyi [skipped: private or protected file] (0ms)
2023-07-20 13:29:19.366 [info] (11248) [IDX(2)]     indexing: <user-path>\.vscode\extensions\ms-python.vscode-pylance-2023.7.30\dist\bundled\stubs\matplotlib\_version.pyi [skipped: private or protected file] (0ms)
2023-07-20 13:29:19.366 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\concurrent\__init__.py [skipped: no '__all__' defined] (0ms)
2023-07-20 13:29:19.366 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\distutils\__init__.py [skipped: no '__all__' defined] (1ms)
2023-07-20 13:29:19.366 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\idlelib\__init__.py [skipped: no '__all__' defined] (0ms)
2023-07-20 13:29:19.367 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\lib2to3\__init__.py [skipped: no '__all__' defined] (0ms)
2023-07-20 13:29:19.367 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\pydoc_data\__init__.py [skipped: no '__all__' defined] (0ms)
2023-07-20 13:29:19.367 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\test\__init__.py [skipped: no '__all__' defined] (1ms)
2023-07-20 13:29:19.367 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\turtledemo\__init__.py [skipped: no '__all__' defined] (0ms)
2023-07-20 13:29:19.367 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\urllib\__init__.py [skipped: no '__all__' defined] (0ms)
2023-07-20 13:29:19.367 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\wsgiref\__init__.py [skipped: no '__all__' defined] (0ms)
2023-07-20 13:29:19.368 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\xmlrpc\__init__.py [skipped: no '__all__' defined] (1ms)
2023-07-20 13:29:19.369 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\site-packages\backcall\__init__.py [skipped: no '__all__' defined] (0ms)
2023-07-20 13:29:19.369 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\site-packages\blackd\__init__.py [skipped: no '__all__' defined] (0ms)
2023-07-20 13:29:19.370 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\site-packages\blib2to3\__init__.py [skipped: no '__all__' defined] (0ms)
2023-07-20 13:29:19.370 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\site-packages\jedi\__init__.py [skipped: no '__all__' defined] (0ms)
2023-07-20 13:29:19.370 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\site-packages\linkify_it\__init__.py [skipped: no '__all__' defined] (0ms)
2023-07-20 13:29:19.371 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\site-packages\matplotlib\pylab.py [skipped: no '__all__' defined] (1ms)
2023-07-20 13:29:19.371 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\site-packages\matplotlib\sphinxext\__init__.py [skipped: no '__all__' defined] (0ms)
2023-07-20 13:29:19.371 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\site-packages\matplotlib\tests\__init__.py [skipped: no '__all__' defined] (0ms)
2023-07-20 13:29:19.371 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\site-packages\matplotlib\tight_bbox.py [skipped: no '__all__' defined] (1ms)
2023-07-20 13:29:19.371 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\site-packages\matplotlib\_animation_data.py [skipped: private or protected file] (0ms)
2023-07-20 13:29:19.371 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\site-packages\matplotlib\_blocking_input.py [skipped: private or protected file] (0ms)
2023-07-20 13:29:19.371 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\site-packages\matplotlib\_cm_listed.py [skipped: private or protected file] (0ms)
2023-07-20 13:29:19.372 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\site-packages\matplotlib\_internal_utils.py [skipped: private or protected file] (0ms)
2023-07-20 13:29:19.372 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\site-packages\matplotlib\_type1font.py [skipped: private or protected file] (0ms)
2023-07-20 13:29:19.372 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\site-packages\matplotlib_inline\__init__.py [skipped: no '__all__' defined] (1ms)
2023-07-20 13:29:19.372 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\site-packages\mpl_toolkits\__init__.py [skipped: no '__all__' defined] (0ms)
2023-07-20 13:29:19.372 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\site-packages\param\__init__.py [skipped: no '__all__' defined] (1ms)
2023-07-20 13:29:19.373 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\site-packages\py4j\__init__.py [skipped: no '__all__' defined] (0ms)
2023-07-20 13:29:19.373 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\site-packages\pyviz_comms\__init__.py [skipped: no '__all__' defined] (0ms)
2023-07-20 13:29:19.373 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\site-packages\seaborn\__init__.py [skipped: no '__all__' defined] (1ms)
2023-07-20 13:29:19.373 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\site-packages\tzdata\__init__.py [skipped: no '__all__' defined] (0ms)
2023-07-20 13:29:19.373 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\site-packages\uc_micro\__init__.py [skipped: no '__all__' defined] (0ms)
2023-07-20 13:29:19.374 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\site-packages\webencodings\__init__.py [skipped: no '__all__' defined] (1ms)
2023-07-20 13:29:19.374 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\site-packages\wheel\__init__.py [skipped: no '__all__' defined] (0ms)
2023-07-20 13:29:19.374 [info] (11248) [IDX(2)]     indexing: <user-path>\mambaforge\envs\py310\Lib\site-packages\xyzservices\__init__.py [skipped: no '__all__' defined] (0ms)
2023-07-20 13:29:19.378 [info] (11248) [IDX(2)]   index execution environment <user-path>\<repo-path> [found 2613 in 111 files] (21ms)
2023-07-20 13:29:19.379 [info] (11248) [IDX(2)] index packages <user-path>\<repo-path> [found 2613 in 1 exec envs] (44ms)
2023-07-20 13:29:19.379 [info] [Info  - 1:29:19 PM] (11248) indexed(2) 111 files over 1 exec env
2023-07-20 13:29:19.401 [info] [Info  - 1:29:19 PM] (11248) Indexing finished(2).
```

</details>

**Comment by ntjess** [comment]
Sorry for all the noise on this issue tracker. After investigating further, I see there is a discrepancy between the `reportUnboundVariable` docs and reality:

Docs:
> The default value for this setting is "error".

Vscode default config is "information":
> ![image](https://github.com/microsoft/pylance-release/assets/23620506/eac7a257-ab62-4899-824f-04b26fd3add3)

Closing since it appears pylance indeed identifies this case, just doesn't show it in the editor without adjusting the default settings.

**Comment by debonte** [comment]
@ntjess, the default value of `reportUnboundVariable` depends on your `python.analysis.typeCheckingMode` setting. If `typeCheckingMode` is `off`, most of our diagnostics rules (including `reportUnboundVariable`) will be set to `none` by default. However, if you change `typeCheckingMode` to `basic` or `strict`, more diagnostics are enabled by default, and `reportUnboundVariable` for example will default to `error`.

I believe that the config setting UI always shows the same default values for our diagnostic rules regardless of your `typeCheckingMode` setting. That's unfortunate, but regardless the default for `reportUnboundVariable` doesn't make sense. I would expect it to be either `none` (the `off` default) or `error` (the `basic` default), but not `information`. 

Seems like we should review these default and make sure they are up-to-date. So I'm reactivating this issue to track that.

**Comment by erictraut** [comment]
The `reportUndefinedVariable` check is performed based entirely on binding information — i.e. information about which symbols are in the symbol table for each scope in a program. The `reportUnboundVariable` check requires code flow analysis and is done as part of type evaluation. That's why it's disabled when type checking is disabled. At least that's the logic behind the current defaults. I think there's an argument that could be made in favor of making `reportUnboundVariable` a "warning" by default when type checking is "off".

**Comment by StellaHuang95** [comment]
@debonte I can go through all the settings and make sure the default values match with our docs.

**Comment by debonte** [comment]
Thanks @StellaHuang95. It's possible that the docs are out of date as well. I believe the source of truth is the `off`/`basic`/`strict` lists in `packages\pyright\packages\pyright-internal\src\common\configOptions.ts` plus the overrides in `_applyPylanceDiagnosticOverrides`.

**Comment by rchiodo** [comment]
This issue has been fixed in prerelease version 2023.7.43, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2023743-31-July-2023-PreRelease)
